File: js/core/protocol/src/e2e/ratchet.ts
import { EncryptionError } from '../utils/errors.js';
import { decryptSessionMessage, encryptSessionMessage } from './messages.js';
import type { LocalDeviceKeyPair, SessionMessage } from './types.js';
import { E2E_PROTOCOL_VERSION } from './types.js';
import {
  bytesToHex,
  diffieHellmanX25519,
  generateX25519KeyPair,
  hexToBytes,
  hkdfSha256,
  randomBytes,
  type X25519KeyPair,
} from './x25519.js';
import type { LocalSessionBootstrapState } from './bootstrap.js';

export const DOUBLE_RATCHET_ROOT_INFO = 'quadra-a/e2e/double-ratchet/root/v1';
export const DOUBLE_RATCHET_CHAIN_INFO = 'quadra-a/e2e/double-ratchet/chain/v1';
export const DOUBLE_RATCHET_CHAIN_SALT = 'quadra-a/e2e/double-ratchet/chain-salt/v1';
export const MAX_SKIPPED_MESSAGE_KEYS = 64;

export interface SkippedMessageKeyState {
  ratchetPublicKey: string;
  messageNumber: number;
  messageKey: string;
}

export interface LocalSessionState {
  sessionId: string;
  peerDid: string;
  peerDeviceId: string;
  selfDeviceId: string;
  role: 'initiator' | 'responder';
  establishedBy: 'prekey';
  phase: 'ratchet-active';
  rootKey: string;
  currentRatchetKey: LocalDeviceKeyPair;
  remoteRatchetPublicKey: string;
  sendingChainKey?: string;
  receivingChainKey?: string;
  createdAt: number;
  updatedAt: number;
  nextSendMessageNumber: number;
  nextReceiveMessageNumber: number;
  previousSendChainLength: number;
  skippedMessageKeys: SkippedMessageKeyState[];
  bootstrap: LocalSessionBootstrapState;
}

export interface CreateRatchetSessionInput {
  sessionId: string;
  peerDid: string;
  peerDeviceId: string;
  selfDeviceId: string;
  role: 'initiator' | 'responder';
  rootKey: Uint8Array;
  currentRatchetKey: X25519KeyPair;
  remoteRatchetPublicKey: Uint8Array;
  bootstrap: LocalSessionBootstrapState;
  createdAt: number;
}

export interface RatchetEncryptInput {
  session: LocalSessionState;
  plaintext: Uint8Array;
  senderDid: string;
  receiverDid: string;
  messageId?: string;
  nonce?: Uint8Array;
  ratchetKeyPair?: X25519KeyPair;
  now?: number;
}

export interface RatchetEncryptResult {
  message: SessionMessage;
  session: LocalSessionState;
  messageKey: Uint8Array;
}

export interface RatchetDecryptInput {
  session: LocalSessionState;
  message: SessionMessage;
  now?: number;
}

export interface RatchetDecryptResult {
  plaintext: Uint8Array;
  session: LocalSessionState;
  messageKey: Uint8Array;
  usedSkippedMessageKey: boolean;
}

function deriveRootAndChainKeys(rootKey: Uint8Array, dhOutput: Uint8Array): {
  rootKey: Uint8Array;
  chainKey: Uint8Array;
} {
  const material = hkdfSha256(dhOutput, rootKey, new TextEncoder().encode(DOUBLE_RATCHET_ROOT_INFO), 64);
  return {
    rootKey: material.slice(0, 32),
    chainKey: material.slice(32, 64),
  };
}

function deriveChainStep(chainKey: Uint8Array): {
  chainKey: Uint8Array;
  messageKey: Uint8Array;
} {
  const material = hkdfSha256(
    chainKey,
    new TextEncoder().encode(DOUBLE_RATCHET_CHAIN_SALT),
    new TextEncoder().encode(DOUBLE_RATCHET_CHAIN_INFO),
    64,
  );
  return {
    chainKey: material.slice(0, 32),
    messageKey: material.slice(32, 64),
  };
}

function cloneSession(session: LocalSessionState): LocalSessionState {
  return {
    ...session,
    currentRatchetKey: { ...session.currentRatchetKey },
    skippedMessageKeys: session.skippedMessageKeys.map((entry) => ({ ...entry })),
  };
}

function generateMessageId(): string {
  return `e2e-msg-${bytesToHex(randomBytes(12))}`;
}

function assertNonce(nonce: Uint8Array): void {
  if (nonce.length !== 24) {
    throw new EncryptionError('Double Ratchet nonce must be 24 bytes', { length: nonce.length });
  }
}

function storeSkippedMessageKey(session: LocalSessionState, ratchetPublicKey: string, messageNumber: number, messageKey: Uint8Array): void {
  session.skippedMessageKeys.push({
    ratchetPublicKey,
    messageNumber,
    messageKey: bytesToHex(messageKey),
  });
  if (session.skippedMessageKeys.length > MAX_SKIPPED_MESSAGE_KEYS) {
    session.skippedMessageKeys.splice(0, session.skippedMessageKeys.length - MAX_SKIPPED_MESSAGE_KEYS);
  }
}

function takeSkippedMessageKey(session: LocalSessionState, ratchetPublicKey: string, messageNumber: number): Uint8Array | null {
  const index = session.skippedMessageKeys.findIndex(
    (entry) => entry.ratchetPublicKey === ratchetPublicKey && entry.messageNumber === messageNumber,
  );
  if (index < 0) {
    return null;
  }
  const [entry] = session.skippedMessageKeys.splice(index, 1);
  return hexToBytes(entry.messageKey);
}

function skipMessageKeys(session: LocalSessionState, untilMessageNumber: number): void {
  if (untilMessageNumber < session.nextReceiveMessageNumber) {
    return;
  }

  if (!session.receivingChainKey) {
    if (untilMessageNumber === session.nextReceiveMessageNumber) {
      return;
    }
    throw new EncryptionError('Missing receiving chain key while skipping messages', {
      untilMessageNumber,
      nextReceiveMessageNumber: session.nextReceiveMessageNumber,
    });
  }

  const gap = untilMessageNumber - session.nextReceiveMessageNumber;
  if (gap > MAX_SKIPPED_MESSAGE_KEYS) {
    throw new EncryptionError('Skipped-message window exceeded', {
      gap,
      maxSkippedMessageKeys: MAX_SKIPPED_MESSAGE_KEYS,
    });
  }

  let chainKey = hexToBytes(session.receivingChainKey);
  while (session.nextReceiveMessageNumber < untilMessageNumber) {
    const step = deriveChainStep(chainKey);
    storeSkippedMessageKey(session, session.remoteRatchetPublicKey, session.nextReceiveMessageNumber, step.messageKey);
    session.nextReceiveMessageNumber += 1;
    chainKey = step.chainKey;
  }
  session.receivingChainKey = bytesToHex(chainKey);
}

function ensureSendingChain(session: LocalSessionState, nextRatchetKeyPair?: X25519KeyPair): void {
  if (session.sendingChainKey) {
    return;
  }

  const ratchetKeyPair = nextRatchetKeyPair ?? generateX25519KeyPair();
  const dhOutput = diffieHellmanX25519(ratchetKeyPair.privateKey, hexToBytes(session.remoteRatchetPublicKey));
  const derived = deriveRootAndChainKeys(hexToBytes(session.rootKey), dhOutput);
  session.rootKey = bytesToHex(derived.rootKey);
  session.currentRatchetKey = {
    publicKey: bytesToHex(ratchetKeyPair.publicKey),
    privateKey: bytesToHex(ratchetKeyPair.privateKey),
  };
  session.sendingChainKey = bytesToHex(derived.chainKey);
  session.previousSendChainLength = session.nextSendMessageNumber;
  session.nextSendMessageNumber = 0;
}

function advanceReceivingRatchet(session: LocalSessionState, remoteRatchetPublicKey: Uint8Array): void {
  const dhOutput = diffieHellmanX25519(
    hexToBytes(session.currentRatchetKey.privateKey),
    remoteRatchetPublicKey,
  );
  const derived = deriveRootAndChainKeys(hexToBytes(session.rootKey), dhOutput);
  session.rootKey = bytesToHex(derived.rootKey);
  session.remoteRatchetPublicKey = bytesToHex(remoteRatchetPublicKey);
  session.receivingChainKey = bytesToHex(derived.chainKey);
  session.nextReceiveMessageNumber = 0;
  session.sendingChainKey = undefined;
}

export function createInitiatorRatchetSession(input: CreateRatchetSessionInput): LocalSessionState {
  const dhOutput = diffieHellmanX25519(input.currentRatchetKey.privateKey, input.remoteRatchetPublicKey);
  const derived = deriveRootAndChainKeys(input.rootKey, dhOutput);
  return {
    sessionId: input.sessionId,
    peerDid: input.peerDid,
    peerDeviceId: input.peerDeviceId,
    selfDeviceId: input.selfDeviceId,
    role: input.role,
    establishedBy: 'prekey',
    phase: 'ratchet-active',
    rootKey: bytesToHex(derived.rootKey),
    currentRatchetKey: {
      publicKey: bytesToHex(input.currentRatchetKey.publicKey),
      privateKey: bytesToHex(input.currentRatchetKey.privateKey),
    },
    remoteRatchetPublicKey: bytesToHex(input.remoteRatchetPublicKey),
    sendingChainKey: bytesToHex(derived.chainKey),
    receivingChainKey: undefined,
    createdAt: input.createdAt,
    updatedAt: input.createdAt,
    nextSendMessageNumber: 0,
    nextReceiveMessageNumber: 0,
    previousSendChainLength: 0,
    skippedMessageKeys: [],
    bootstrap: input.bootstrap,
  };
}

export function createResponderRatchetSession(input: CreateRatchetSessionInput): LocalSessionState {
  const dhOutput = diffieHellmanX25519(input.currentRatchetKey.privateKey, input.remoteRatchetPublicKey);
  const derived = deriveRootAndChainKeys(input.rootKey, dhOutput);
  return {
    sessionId: input.sessionId,
    peerDid: input.peerDid,
    peerDeviceId: input.peerDeviceId,
    selfDeviceId: input.selfDeviceId,
    role: input.role,
    establishedBy: 'prekey',
    phase: 'ratchet-active',
    rootKey: bytesToHex(derived.rootKey),
    currentRatchetKey: {
      publicKey: bytesToHex(input.currentRatchetKey.publicKey),
      privateKey: bytesToHex(input.currentRatchetKey.privateKey),
    },
    remoteRatchetPublicKey: bytesToHex(input.remoteRatchetPublicKey),
    sendingChainKey: undefined,
    receivingChainKey: bytesToHex(derived.chainKey),
    createdAt: input.createdAt,
    updatedAt: input.createdAt,
    nextSendMessageNumber: 0,
    nextReceiveMessageNumber: 0,
    previousSendChainLength: 0,
    skippedMessageKeys: [],
    bootstrap: input.bootstrap,
  };
}

export function encryptRatchetMessage(input: RatchetEncryptInput): RatchetEncryptResult {
  const session = cloneSession(input.session);
  ensureSendingChain(session, input.ratchetKeyPair);
  if (!session.sendingChainKey) {
    throw new EncryptionError('Missing sending chain key after ratchet initialization');
  }

  const chainStep = deriveChainStep(hexToBytes(session.sendingChainKey));
  const nonce = input.nonce ?? randomBytes(24);
  assertNonce(nonce);
  const message = encryptSessionMessage({
    version: E2E_PROTOCOL_VERSION,
    type: 'SESSION_MESSAGE',
    senderDid: input.senderDid,
    receiverDid: input.receiverDid,
    senderDeviceId: session.selfDeviceId,
    receiverDeviceId: session.peerDeviceId,
    sessionId: session.sessionId,
    messageId: input.messageId ?? generateMessageId(),
    ratchetPublicKey: hexToBytes(session.currentRatchetKey.publicKey),
    previousChainLength: session.previousSendChainLength,
    messageNumber: session.nextSendMessageNumber,
    nonce,
  }, chainStep.messageKey, input.plaintext);

  session.sendingChainKey = bytesToHex(chainStep.chainKey);
  session.nextSendMessageNumber += 1;
  session.updatedAt = input.now ?? Date.now();

  return {
    message,
    session,
    messageKey: chainStep.messageKey,
  };
}

export function decryptRatchetMessage(input: RatchetDecryptInput): RatchetDecryptResult {
  const session = cloneSession(input.session);
  const skippedMessageKey = takeSkippedMessageKey(
    session,
    bytesToHex(input.message.ratchetPublicKey),
    input.message.messageNumber,
  );
  if (skippedMessageKey) {
    const plaintext = decryptSessionMessage(input.message, skippedMessageKey);
    session.updatedAt = input.now ?? Date.now();
    return {
      plaintext,
      session,
      messageKey: skippedMessageKey,
      usedSkippedMessageKey: true,
    };
  }

  if (bytesToHex(input.message.ratchetPublicKey) !== session.remoteRatchetPublicKey) {
    skipMessageKeys(session, input.message.previousChainLength);
    advanceReceivingRatchet(session, input.message.ratchetPublicKey);
  }

  skipMessageKeys(session, input.message.messageNumber);
  if (!session.receivingChainKey) {
    throw new EncryptionError('Missing receiving chain key for session message decryption');
  }

  const chainStep = deriveChainStep(hexToBytes(session.receivingChainKey));
  const plaintext = decryptSessionMessage(input.message, chainStep.messageKey);
  session.receivingChainKey = bytesToHex(chainStep.chainKey);
  session.nextReceiveMessageNumber = input.message.messageNumber + 1;
  session.updatedAt = input.now ?? Date.now();

  return {
    plaintext,
    session,
    messageKey: chainStep.messageKey,
    usedSkippedMessageKey: false,
  };
}

File: rust/core/src/e2e/ratchet.rs
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};

use crate::e2e::bootstrap::LocalSessionBootstrapState;
use crate::e2e::crypto::{
    bytes_to_hex, diffie_hellman_x25519, generate_x25519_key_pair, hkdf_sha256, random_bytes,
    X25519KeyPair,
};
use crate::e2e::messages::{decrypt_session_message, encrypt_session_message};
use crate::e2e::types::{E2EMessageType, LocalDeviceKeyPair, SessionMessage, E2E_PROTOCOL_VERSION};

pub const DOUBLE_RATCHET_ROOT_INFO: &str = "quadra-a/e2e/double-ratchet/root/v1";
pub const DOUBLE_RATCHET_CHAIN_INFO: &str = "quadra-a/e2e/double-ratchet/chain/v1";
pub const DOUBLE_RATCHET_CHAIN_SALT: &str = "quadra-a/e2e/double-ratchet/chain-salt/v1";
pub const MAX_SKIPPED_MESSAGE_KEYS: usize = 64;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct SkippedMessageKeyState {
    #[serde(rename = "ratchetPublicKey")]
    pub ratchet_public_key: String,
    #[serde(rename = "messageNumber")]
    pub message_number: u32,
    #[serde(rename = "messageKey")]
    pub message_key: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LocalSessionState {
    #[serde(rename = "sessionId")]
    pub session_id: String,
    #[serde(rename = "peerDid")]
    pub peer_did: String,
    #[serde(rename = "peerDeviceId")]
    pub peer_device_id: String,
    #[serde(rename = "selfDeviceId")]
    pub self_device_id: String,
    pub role: String,
    #[serde(rename = "establishedBy")]
    pub established_by: String,
    pub phase: String,
    #[serde(rename = "rootKey")]
    pub root_key: String,
    #[serde(rename = "currentRatchetKey")]
    pub current_ratchet_key: LocalDeviceKeyPair,
    #[serde(rename = "remoteRatchetPublicKey")]
    pub remote_ratchet_public_key: String,
    #[serde(
        default,
        rename = "sendingChainKey",
        skip_serializing_if = "Option::is_none"
    )]
    pub sending_chain_key: Option<String>,
    #[serde(
        default,
        rename = "receivingChainKey",
        skip_serializing_if = "Option::is_none"
    )]
    pub receiving_chain_key: Option<String>,
    #[serde(rename = "createdAt")]
    pub created_at: u64,
    #[serde(rename = "updatedAt")]
    pub updated_at: u64,
    #[serde(rename = "nextSendMessageNumber")]
    pub next_send_message_number: u32,
    #[serde(rename = "nextReceiveMessageNumber")]
    pub next_receive_message_number: u32,
    #[serde(rename = "previousSendChainLength")]
    pub previous_send_chain_length: u32,
    #[serde(rename = "skippedMessageKeys")]
    pub skipped_message_keys: Vec<SkippedMessageKeyState>,
    pub bootstrap: LocalSessionBootstrapState,
}

pub struct CreateRatchetSessionInput {
    pub session_id: String,
    pub peer_did: String,
    pub peer_device_id: String,
    pub self_device_id: String,
    pub role: String,
    pub root_key: Vec<u8>,
    pub current_ratchet_key: X25519KeyPair,
    pub remote_ratchet_public_key: Vec<u8>,
    pub bootstrap: LocalSessionBootstrapState,
    pub created_at: u64,
}

pub struct RatchetEncryptInput<'a> {
    pub session: &'a LocalSessionState,
    pub plaintext: &'a [u8],
    pub sender_did: &'a str,
    pub receiver_did: &'a str,
    pub message_id: Option<String>,
    pub nonce: Option<Vec<u8>>,
    pub ratchet_keypair: Option<X25519KeyPair>,
    pub now: Option<u64>,
}

pub struct RatchetEncryptResult {
    pub message: SessionMessage,
    pub session: LocalSessionState,
    pub message_key: Vec<u8>,
}

pub struct RatchetDecryptInput<'a> {
    pub session: &'a LocalSessionState,
    pub message: &'a SessionMessage,
    pub now: Option<u64>,
}

pub struct RatchetDecryptResult {
    pub plaintext: Vec<u8>,
    pub session: LocalSessionState,
    pub message_key: Vec<u8>,
    pub used_skipped_message_key: bool,
}

fn derive_root_and_chain_keys(root_key: &[u8], dh_output: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
    let material = hkdf_sha256(dh_output, root_key, DOUBLE_RATCHET_ROOT_INFO.as_bytes(), 64)?;
    Ok((material[0..32].to_vec(), material[32..64].to_vec()))
}

fn derive_chain_step(chain_key: &[u8]) -> Result<(Vec<u8>, Vec<u8>)> {
    let material = hkdf_sha256(
        chain_key,
        DOUBLE_RATCHET_CHAIN_SALT.as_bytes(),
        DOUBLE_RATCHET_CHAIN_INFO.as_bytes(),
        64,
    )?;
    Ok((material[0..32].to_vec(), material[32..64].to_vec()))
}

fn decode_hex_key(value: &str, label: &str) -> Result<Vec<u8>> {
    hex::decode(value).map_err(|_| anyhow!("Failed to decode {} hex", label))
}

fn generate_message_id() -> String {
    format!("e2e-msg-{}", hex::encode(random_bytes(12)))
}

fn assert_nonce(nonce: &[u8]) -> Result<()> {
    if nonce.len() != 24 {
        anyhow::bail!("Double Ratchet nonce must be 24 bytes, got {}", nonce.len());
    }
    Ok(())
}

fn store_skipped_message_key(
    session: &mut LocalSessionState,
    ratchet_public_key: &str,
    message_number: u32,
    message_key: &[u8],
) {
    session.skipped_message_keys.push(SkippedMessageKeyState {
        ratchet_public_key: ratchet_public_key.to_string(),
        message_number,
        message_key: bytes_to_hex(message_key),
    });
    if session.skipped_message_keys.len() > MAX_SKIPPED_MESSAGE_KEYS {
        let overflow = session.skipped_message_keys.len() - MAX_SKIPPED_MESSAGE_KEYS;
        session.skipped_message_keys.drain(0..overflow);
    }
}

fn take_skipped_message_key(
    session: &mut LocalSessionState,
    ratchet_public_key: &str,
    message_number: u32,
) -> Result<Option<Vec<u8>>> {
    let Some(index) = session.skipped_message_keys.iter().position(|entry| {
        entry.ratchet_public_key == ratchet_public_key && entry.message_number == message_number
    }) else {
        return Ok(None);
    };

    let entry = session.skipped_message_keys.remove(index);
    Ok(Some(decode_hex_key(
        &entry.message_key,
        "skipped message key",
    )?))
}

fn skip_message_keys(session: &mut LocalSessionState, until_message_number: u32) -> Result<()> {
    if until_message_number < session.next_receive_message_number {
        return Ok(());
    }

    let Some(mut chain_key) = session
        .receiving_chain_key
        .as_ref()
        .map(|value| decode_hex_key(value, "receiving chain key"))
        .transpose()?
    else {
        if until_message_number == session.next_receive_message_number {
            return Ok(());
        }
        anyhow::bail!(
            "Missing receiving chain key while skipping messages: until={}, next={}",
            until_message_number,
            session.next_receive_message_number
        );
    };

    let gap = until_message_number - session.next_receive_message_number;
    if gap as usize > MAX_SKIPPED_MESSAGE_KEYS {
        anyhow::bail!(
            "Skipped-message window exceeded: gap={}, max={}",
            gap,
            MAX_SKIPPED_MESSAGE_KEYS
        );
    }

    while session.next_receive_message_number < until_message_number {
        let (next_chain_key, message_key) = derive_chain_step(&chain_key)?;
        let remote_ratchet_public_key = session.remote_ratchet_public_key.clone();
        let skipped_message_number = session.next_receive_message_number;
        store_skipped_message_key(
            session,
            &remote_ratchet_public_key,
            skipped_message_number,
            &message_key,
        );
        session.next_receive_message_number += 1;
        chain_key = next_chain_key;
    }

    session.receiving_chain_key = Some(bytes_to_hex(&chain_key));
    Ok(())
}

fn ensure_sending_chain(
    session: &mut LocalSessionState,
    next_ratchet_keypair: Option<X25519KeyPair>,
) -> Result<()> {
    if session.sending_chain_key.is_some() {
        return Ok(());
    }

    let ratchet_keypair = next_ratchet_keypair.unwrap_or_else(generate_x25519_key_pair);
    let dh_output = diffie_hellman_x25519(
        &ratchet_keypair.private_key,
        &decode_hex_key(
            &session.remote_ratchet_public_key,
            "remote ratchet public key",
        )?,
    )?;
    let (root_key, chain_key) =
        derive_root_and_chain_keys(&decode_hex_key(&session.root_key, "root key")?, &dh_output)?;
    session.root_key = bytes_to_hex(&root_key);
    session.current_ratchet_key = LocalDeviceKeyPair {
        public_key: bytes_to_hex(&ratchet_keypair.public_key),
        private_key: bytes_to_hex(&ratchet_keypair.private_key),
    };
    session.sending_chain_key = Some(bytes_to_hex(&chain_key));
    session.previous_send_chain_length = session.next_send_message_number;
    session.next_send_message_number = 0;
    Ok(())
}

fn advance_receiving_ratchet(
    session: &mut LocalSessionState,
    remote_ratchet_public_key: &[u8],
) -> Result<()> {
    let dh_output = diffie_hellman_x25519(
        &decode_hex_key(
            &session.current_ratchet_key.private_key,
            "current ratchet private key",
        )?,
        remote_ratchet_public_key,
    )?;
    let (root_key, chain_key) =
        derive_root_and_chain_keys(&decode_hex_key(&session.root_key, "root key")?, &dh_output)?;
    session.root_key = bytes_to_hex(&root_key);
    session.remote_ratchet_public_key = bytes_to_hex(remote_ratchet_public_key);
    session.receiving_chain_key = Some(bytes_to_hex(&chain_key));
    session.next_receive_message_number = 0;
    session.sending_chain_key = None;
    Ok(())
}

pub fn create_initiator_ratchet_session(
    input: CreateRatchetSessionInput,
) -> Result<LocalSessionState> {
    let dh_output = diffie_hellman_x25519(
        &input.current_ratchet_key.private_key,
        &input.remote_ratchet_public_key,
    )?;
    let (root_key, chain_key) = derive_root_and_chain_keys(&input.root_key, &dh_output)?;
    Ok(LocalSessionState {
        session_id: input.session_id,
        peer_did: input.peer_did,
        peer_device_id: input.peer_device_id,
        self_device_id: input.self_device_id,
        role: input.role,
        established_by: "prekey".to_string(),
        phase: "ratchet-active".to_string(),
        root_key: bytes_to_hex(&root_key),
        current_ratchet_key: LocalDeviceKeyPair {
            public_key: bytes_to_hex(&input.current_ratchet_key.public_key),
            private_key: bytes_to_hex(&input.current_ratchet_key.private_key),
        },
        remote_ratchet_public_key: bytes_to_hex(&input.remote_ratchet_public_key),
        sending_chain_key: Some(bytes_to_hex(&chain_key)),
        receiving_chain_key: None,
        created_at: input.created_at,
        updated_at: input.created_at,
        next_send_message_number: 0,
        next_receive_message_number: 0,
        previous_send_chain_length: 0,
        skipped_message_keys: Vec::new(),
        bootstrap: input.bootstrap,
    })
}

pub fn create_responder_ratchet_session(
    input: CreateRatchetSessionInput,
) -> Result<LocalSessionState> {
    let dh_output = diffie_hellman_x25519(
        &input.current_ratchet_key.private_key,
        &input.remote_ratchet_public_key,
    )?;
    let (root_key, chain_key) = derive_root_and_chain_keys(&input.root_key, &dh_output)?;
    Ok(LocalSessionState {
        session_id: input.session_id,
        peer_did: input.peer_did,
        peer_device_id: input.peer_device_id,
        self_device_id: input.self_device_id,
        role: input.role,
        established_by: "prekey".to_string(),
        phase: "ratchet-active".to_string(),
        root_key: bytes_to_hex(&root_key),
        current_ratchet_key: LocalDeviceKeyPair {
            public_key: bytes_to_hex(&input.current_ratchet_key.public_key),
            private_key: bytes_to_hex(&input.current_ratchet_key.private_key),
        },
        remote_ratchet_public_key: bytes_to_hex(&input.remote_ratchet_public_key),
        sending_chain_key: None,
        receiving_chain_key: Some(bytes_to_hex(&chain_key)),
        created_at: input.created_at,
        updated_at: input.created_at,
        next_send_message_number: 0,
        next_receive_message_number: 0,
        previous_send_chain_length: 0,
        skipped_message_keys: Vec::new(),
        bootstrap: input.bootstrap,
    })
}

pub fn encrypt_ratchet_message(input: RatchetEncryptInput<'_>) -> Result<RatchetEncryptResult> {
    let mut session = input.session.clone();
    ensure_sending_chain(&mut session, input.ratchet_keypair)?;

    let sending_chain_key = session
        .sending_chain_key
        .as_ref()
        .ok_or_else(|| anyhow!("Missing sending chain key after ratchet initialization"))?;
    let (next_chain_key, message_key) =
        derive_chain_step(&decode_hex_key(sending_chain_key, "sending chain key")?)?;
    let nonce = input.nonce.unwrap_or_else(|| random_bytes(24));
    assert_nonce(&nonce)?;
    let message = encrypt_session_message(
        &SessionMessage {
            version: E2E_PROTOCOL_VERSION,
            message_type: E2EMessageType::SessionMessage,
            sender_did: input.sender_did.to_string(),
            receiver_did: input.receiver_did.to_string(),
            sender_device_id: session.self_device_id.clone(),
            receiver_device_id: session.peer_device_id.clone(),
            session_id: session.session_id.clone(),
            message_id: input.message_id.unwrap_or_else(generate_message_id),
            ratchet_public_key: decode_hex_key(
                &session.current_ratchet_key.public_key,
                "current ratchet public key",
            )?,
            previous_chain_length: session.previous_send_chain_length,
            message_number: session.next_send_message_number,
            nonce,
            ciphertext: Vec::new(),
        },
        &message_key,
        input.plaintext,
    )?;

    session.sending_chain_key = Some(bytes_to_hex(&next_chain_key));
    session.next_send_message_number += 1;
    session.updated_at = input.now.unwrap_or(session.updated_at);

    Ok(RatchetEncryptResult {
        message,
        session,
        message_key,
    })
}

pub fn decrypt_ratchet_message(input: RatchetDecryptInput<'_>) -> Result<RatchetDecryptResult> {
    let mut session = input.session.clone();
    let ratchet_public_key = bytes_to_hex(&input.message.ratchet_public_key);

    if let Some(message_key) = take_skipped_message_key(
        &mut session,
        &ratchet_public_key,
        input.message.message_number,
    )? {
        let plaintext = decrypt_session_message(input.message, &message_key)?;
        session.updated_at = input.now.unwrap_or(session.updated_at);
        return Ok(RatchetDecryptResult {
            plaintext,
            session,
            message_key,
            used_skipped_message_key: true,
        });
    }

    if ratchet_public_key != session.remote_ratchet_public_key {
        skip_message_keys(&mut session, input.message.previous_chain_length)?;
        advance_receiving_ratchet(&mut session, &input.message.ratchet_public_key)?;
    }

    skip_message_keys(&mut session, input.message.message_number)?;
    let receiving_chain_key = session
        .receiving_chain_key
        .as_ref()
        .ok_or_else(|| anyhow!("Missing receiving chain key for session message decryption"))?;
    let (next_chain_key, message_key) =
        derive_chain_step(&decode_hex_key(receiving_chain_key, "receiving chain key")?)?;
    let plaintext = decrypt_session_message(input.message, &message_key)?;
    session.receiving_chain_key = Some(bytes_to_hex(&next_chain_key));
    session.next_receive_message_number = input.message.message_number + 1;
    session.updated_at = input.now.unwrap_or(session.updated_at);

    Ok(RatchetDecryptResult {
        plaintext,
        session,
        message_key,
        used_skipped_message_key: false,
    })
}
